opm/autodiff/SimulatorBase_impl.hpp

```diff
@@ -667,21 +671,27 @@ namespace Opm
     SimulatorBase<Implementation>::FIPUnitConvert(const UnitSystem& units,
                                                   std::vector<V>& fip)
     {
+        for (size_t i = 0; i < fip.size(); ++i) {
+            FIPUnitConvert(units, fip[i]);
+        }
+    }
+
+
+    template <class Implementation>
+    void
+    SimulatorBase<Implementation>::FIPUnitConvert(const UnitSystem& units, V& fip)
+    {
         if (units.getType() == UnitSystem::UnitType::UNIT_TYPE_FIELD) {
-            for (size_t i = 0; i < fip.size(); ++i) {
-                fip[i][0] = unit::convert::to(fip[i][0], unit::stb);
-                fip[i][1] = unit::convert::to(fip[i][1], unit::stb); 
-                fip[i][2] = unit::convert::to(fip[i][2], 1000*unit::cubic(unit::feet));
-                fip[i][3] = unit::convert::to(fip[i][3], 1000*unit::cubic(unit::feet));
-                fip[i][4] = unit::convert::to(fip[i][4], unit::stb);
-                fip[i][5] = unit::convert::to(fip[i][5], unit::stb);
-                fip[i][6] = unit::convert::to(fip[i][6], unit::psia);
-            }
+            fip[0] = unit::convert::to(fip[0], unit::stb);
+            fip[1] = unit::convert::to(fip[1], unit::stb);
+            fip[2] = unit::convert::to(fip[2], 1000*unit::cubic(unit::feet));
+            fip[3] = unit::convert::to(fip[3], 1000*unit::cubic(unit::feet));
+            fip[4] = unit::convert::to(fip[4], unit::stb);
+            fip[5] = unit::convert::to(fip[5], unit::stb);
+            fip[6] = unit::convert::to(fip[6], unit::psia);
         }
-        if (units.getType() == UnitSystem::UnitType::UNIT_TYPE_METRIC) {
-            for (size_t i = 0; i < fip.size(); ++i) {
-                fip[i][6] = unit::convert::to(fip[i][6], unit::barsa);
-            }
+        else if (units.getType() == UnitSystem::UnitType::UNIT_TYPE_METRIC) {
+            fip[6] = unit::convert::to(fip[6], unit::barsa);
         }
```

Comment:
To guard agains future trouble, we should add a final else branch which throws (in case someone tries to run with LAB units).